test/tests/TimerTest.cpp

```diff
@@ -0,0 +1,279 @@
+
+    // called by Timer::InitializeIARM
+    EXPECT_CALL(iarmBusImplMock, IARM_Bus_Connect())
+        .Times(1)
+        .WillOnce(::testing::Return(IARM_RESULT_SUCCESS));
+
+    /////////////// IARM expectations - End //////////////////////////////
+
+    EXPECT_CALL(service, Submit(::testing::_, ::testing::_))
+        .Times(2)
+        // called by Timer::sendTimerExpiryReminder
+        .WillOnce(::testing::Invoke(
+            [&](const uint32_t, const Core::ProxyType<Core::JSON::IElement>& json) {
+                string text;
+                EXPECT_TRUE(json->ToString(text));
+                EXPECT_EQ(text, string(_T("{"
+                                          "\"jsonrpc\":\"2.0\","
+                                          "\"method\":\"org.rdk.Timer.timerExpiryReminder\","
+                                          "\"params\":{\"timerId\":0,"
+                                           "\"mode\":\"WAKE\","
+                                           "\"timeRemaining\":0}"
+                                          "}")));
+
+                return Core::ERROR_NONE;
+            }))
+    
+        // called by Timer::sendTimerExpired
+        .WillOnce(::testing::Invoke(
+            [&](const uint32_t, const Core::ProxyType<Core::JSON::IElement>& json) {
+                string text;
+                EXPECT_TRUE(json->ToString(text));
+                EXPECT_EQ(text, string(_T("{"
+                                          "\"jsonrpc\":\"2.0\","
+                                          "\"method\":\"org.rdk.Timer.timerExpired\","
+                                          "\"params\":{\"timerId\":0,"
+                                           "\"mode\":\"WAKE\","
+                                           "\"status\":0}"
+                                          "}")));
+
+                return Core::ERROR_NONE;
+            }));
+
+    // Initialize
+    EXPECT_EQ(string(""), plugin->Initialize(nullptr));
+
+    //Create a timer of 0.2 sec with a reminder of 0.1 sec
+    EXPECT_EQ(Core::ERROR_NONE, handler.Invoke(connection, _T("startTimer"), 
+                        _T("{\"interval\":0.2, \"mode\":\"WAKE\", \"remindBefore\": 0.1}"), response));
+    EXPECT_THAT(response, testing::HasSubstr("timerId"));
+    EXPECT_THAT(response, testing::HasSubstr("\"success\":true"));
+
+    //Extract timer id from response
+    JsonObject params;
+    EXPECT_TRUE(params.FromString(response));
+    EXPECT_TRUE(params.HasLabel(_T("timerId")));
+
+    // JSON-RPC events
+    auto dispatcher = static_cast<PluginHost::IDispatcher*>(
+        plugin->QueryInterface(PluginHost::IDispatcher::ID));
+    EXPECT_TRUE(dispatcher != nullptr);
+
+    dispatcher->Activate(&service);
+
+    handler.Subscribe(0, _T("timerExpiryReminder"), _T("org.rdk.Timer"), message);
+    handler.Subscribe(0, _T("timerExpired"), _T("org.rdk.Timer"), message);
+
+    //Wait for timer expiry & reminder calls
+    Core::Event wait(false, true);
+    wait.Lock(200);
```

Comment:
Activate and Subscribe should be before "startTimer" otherwise it's a race condition. You need 2 events, one for "timerExpiryReminder" and one for "timerExpired". "Lock" should be a safe timeout, like 2 seconds otherwise it's a race condition. Please use SetEvent() to stop locks and check "Lock" returns Core::ERROR_NONE.

Reply:
I don't understand how SetEvent() & Lock() works. Can you provide more info or the sample code for this comment, "Please use SetEvent() to stop locks and check "Lock" returns Core::ERROR_NONE"?